drivers/gpio/gpio_andes_atcgpio100.c

```diff
@@ -0,0 +1,370 @@
+#define REG_IMD3 0x60 /* Interrupt mode 24 ~ 31 reg.    */
+#define REG_ISTA 0x64 /* Interrupt status reg.          */
+#define REG_DEBE 0x70 /* De-bounce enable reg.          */
+#define REG_DEBC 0x74 /* De-Bounce control reg.         */
+
+#define INT_NO_OPERATION        0x0
+#define INT_HIGH_LEVEL          0x2
+#define INT_LOW_LEVEL           0x3
+#define INT_NEGATIVE_EDGE       0x5
+#define INT_POSITIVE_EDGE       0x6
+#define INT_DUAL_EDGE           0x7
+
+#define PULL_CONFIGURED         (1 << 31)
+#define DEBOUNCE_CONFIGURED     (1 << 29)
+#define DF_DEBOUNCED_SETTING    (0x80000003)
+
+#define DEV_CFG(dev) \
+	((const struct gpio_atcgpio100_config * const)(dev)->config)
+
+#define DEV_DATA(dev) \
+	((struct gpio_atcgpio100_data *)(dev)->data)
+
+
+#define GPIO_CFG(dev)  (DEV_CFG(dev)->base + REG_CFG)
+#define GPIO_DIR(dev)  (DEV_CFG(dev)->base + REG_DIR)
+#define GPIO_DIN(dev)  (DEV_CFG(dev)->base + REG_DIN)
+#define GPIO_DOUT(dev) (DEV_CFG(dev)->base + REG_DOUT)
+#define GPIO_DCLR(dev) (DEV_CFG(dev)->base + REG_DCLR)
+#define GPIO_DSET(dev) (DEV_CFG(dev)->base + REG_DSET)
+#define GPIO_PUEN(dev) (DEV_CFG(dev)->base + REG_PUEN)
+#define GPIO_PTYP(dev) (DEV_CFG(dev)->base + REG_PTYP)
+#define GPIO_INTE(dev) (DEV_CFG(dev)->base + REG_INTE)
+#define GPIO_IMD(dev, idx) (DEV_CFG(dev)->base + REG_IMD0 + (idx << 2))
+#define GPIO_ISTA(dev) (DEV_CFG(dev)->base + REG_ISTA)
+#define GPIO_DEBE(dev) (DEV_CFG(dev)->base + REG_DEBE)
+#define GPIO_DEBC(dev) (DEV_CFG(dev)->base + REG_DEBC)
+
+#define INWORD(x)     sys_read32(x)
+#define OUTWORD(x, d) sys_write32(d, x)
+
+#define SET_GPIO_INT_MODE(cur_val, mode, ch_idx)		\
+		do {						\
+			cur_val &= ~(0x7 << (ch_idx << 2));	\
+			cur_val |= (mode << (ch_idx << 2));	\
+		} while (false)
+
+
+typedef void (*atcgpio100_cfg_func_t)(void);
+
+struct gpio_atcgpio100_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	uint32_t                  base;
+	uint32_t                  irq_num;
+	atcgpio100_cfg_func_t	  cfg_func;
+};
+
+struct gpio_atcgpio100_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	/* list of callbacks */
+	sys_slist_t cb;
+	struct k_spinlock lock;
+};
+
+static int gpio_atcgpio100_config(const struct device *port,
+				  gpio_pin_t pin,
+				  gpio_flags_t flags)
+{
+	struct gpio_atcgpio100_data * const data = DEV_DATA(port);
+	uint32_t port_value, pin_mask, io_flags;
+	k_spinlock_key_t key;
+
+	/* Does not support disconnected pin, and
+	 * not supporting both input/output at same time.
+	 */
+	io_flags = flags & (GPIO_INPUT | GPIO_OUTPUT);
+	if ((io_flags == GPIO_DISCONNECTED)
+	    || (io_flags == (GPIO_INPUT | GPIO_OUTPUT))) {
+		return -ENOTSUP;
+	}
+
+	pin_mask = (1 << pin);
+
+	if (flags & GPIO_OUTPUT) {
+
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			OUTWORD(GPIO_DSET(port), pin_mask);
+		}
+		if (flags & GPIO_OUTPUT_INIT_LOW) {
+			OUTWORD(GPIO_DCLR(port), pin_mask);
+		}
```

Comment:
Can a single pin be set to both `INIT_HIGH` and `INIT_LOW`? If not, this probably should be in a if/else if/else block.

Reply:
They can't be set at same time. Using else if will make it more clearly.
Done.